bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java

```diff
@@ -486,7 +486,13 @@ public Image(Device device, ImageData source, ImageData mask) {
 public Image (Device device, InputStream stream) {
 	super(device);
 	if (stream == null) SWT.error(SWT.ERROR_NULL_ARGUMENT);
-	this.imageProvider = new ImageDataLoaderStreamProviderWrapper(stream);
+	if(ImageDataLoader.isDynamicallySizable(stream))
+	{
+		this.imageProvider = new DynamicallySizableImageDataLoaderStreamProviderWrapper(stream);
+	}
+	else {
+		this.imageProvider = new ImageDataLoaderStreamProviderWrapper(stream);
+	}
```

Comment:
When integrating a preliminary versin of this into our product, I found that this implementation can easily break: `ImageDataLoader.isDynamicallySizable(stream)` may already read the stream such that the created wrapper cannot read it anymore. That's why the wrapper initially reads the whole stream and stores it internally to then create a `ByteArrayInputStream` out of it on demand. We probably need to do something similar here. Maybe we could even use the same wrapper for both cases? Actually, `DynamicallySizableImageDataLoaderSTreamProviderWrapper` has an additional `intpuStream` field, which is also used to load an image at size, but that field is never initialized. How does that work at all?